python/src/numerous/collection/numerous_collection.py

```diff
@@ -37,12 +37,82 @@ def document(self, key: str) -> NumerousDocument:
         """
         numerous_doc_ref = self._client.get_collection_document(self.key, key)
         if numerous_doc_ref is not None:
-
             numerous_document = NumerousDocument(
-                self._client, numerous_doc_ref.key, self.id, numerous_doc_ref
+                self._client,
+                numerous_doc_ref.key,
+                (self.id, self.key),
+                numerous_doc_ref,
             )
         else:
-            numerous_document = NumerousDocument(self._client, key, self.id)
+            numerous_document = NumerousDocument(self._client, key, (self.id, self.key))
 
-        self.documents.update({numerous_document.key: numerous_document})
         return numerous_document
+
+    def documents(
+        self, tag_key: Optional[str] = None, tag_value: Optional[str] = None
+    ) -> Iterator[NumerousDocument]:
+        """
+        Retrieve documents from the collection, filtered by a tag key and value.
+
+        Parameters
+        ----------
+        tag_key : Optional[str]
+            The key of the tag used to filter documents (optional).
+        tag_value : Optional[str]
+            The value of the tag used to filter documents (optional).
+
+        Yields
+        ------
+        NumerousDocument
+            Yields NumerousDocument objects from the collection.
+
+        """
+        end_cursor = ""
+        while True:
+            tag_input = None
+            if tag_key is not None and tag_value is not None:
+                tag_input = TagInput(key=tag_key, value=tag_value)
+            result = self._client.get_collection_documents(
+                self.key, end_cursor, tag_input
+            )
+            if result is None:
+                break
+            numerous_doc_refs, has_next_page, end_cursor = result
+            if numerous_doc_refs is None:
+                break
+            for numerous_doc_ref in numerous_doc_refs:
+                if numerous_doc_ref is None:
+                    return
+                yield NumerousDocument(
+                    self._client,
+                    numerous_doc_ref.key,
+                    (self.id, self.key),
+                    numerous_doc_ref,
+                )
+            if not has_next_page:
+                break
```

Comment:
I think some parts can be taken out of the loop, and some parts simplified a bit. Similar changes can be made for collections. Have a look at this:

```suggestion
        end_cursor = ""

        tag_input = None
        if tag_key is not None and tag_value is not None:
            tag_input = TagInput(key=tag_key, value=tag_value)

        has_next_page = True
        while has_next_page:
            result = self._client.get_collection_documents(
                self.key, end_cursor, tag_input
            )

            if result is None:
                break

            numerous_doc_refs, has_next_page, end_cursor = result
            if numerous_doc_refs is None:
                break

            for numerous_doc_ref in numerous_doc_refs:
                if numerous_doc_ref is None:
                    continue
                yield NumerousDocument(
                    self._client,
                    numerous_doc_ref.key,
                    (self.id, self.key),
                    numerous_doc_ref,
                )

```